src/libraries/System.Text.Json/ref/System.Text.Json.cs

```diff
@@ -1040,10 +1040,11 @@ public sealed partial class JsonSourceGenerationOptionsAttribute : System.Text.J
         public System.Text.Json.Serialization.JsonKnownNamingPolicy PropertyNamingPolicy { get { throw null; } set { } }
         public bool WriteIndented { get { throw null; } set { } }
     }
-    [System.Diagnostics.CodeAnalysis.RequiresDynamicCodeAttribute("JsonStringEnumConverter cannot be statically analyzed and requires runtime code generation. Consider authoring a custom converter that is not a factory to work around the issue. See https://github.com/dotnet/runtime/issues/73124.")]
     public partial class JsonStringEnumConverter : System.Text.Json.Serialization.JsonConverterFactory
     {
+        [System.Diagnostics.CodeAnalysis.RequiresDynamicCodeAttribute("JsonStringEnumConverter cannot be statically analyzed and requires runtime code generation. Native AOT users should only use this type indirectly via JsonConverterAttribute annotations.")]
```

Comment:
We should test this out, but I _think_ this is still going to give warnings when being used in AOT apps. JsonConverterAttribute is annotated with `[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)]`:

https://github.com/dotnet/runtime/blob/2bf8f1aa83e192a307d5846424880cd61bec1a4f/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Attributes/JsonConverterAttribute.cs#L25

which means it is going to reflect against the parameterless constructor of the Type being specified. Here we are marking that constructor as "RequiresDynamicCode". The AOT compiler is going to warn that there is reflection being done against a constructor marked as "unsafe for AOT".

cc @MichalStrehovsky @vitek-karas

Reply:
FWIW I tested this code:
```C#
using System;
using System.Reflection;
using System.Text.Json;
using System.Text.Json.Serialization;

Console.WriteLine(JsonSerializer.Serialize(new MyPoco(), MyContext.Default.MyPoco));

public class MyPoco
{
    [JsonConverter(typeof(JsonStringEnumConverter))]
    public BindingFlags Flags { get; set; }
}

[JsonSerializable(typeof(MyPoco))]
public partial class MyContext : JsonSerializerContext
{
}
```
and it appears to not be warning on the attribute annotation location. The concern though is that future revisions of the AOT compiler could be warning there.

Reply:
@vitek-karas pointed out that the above not warning is accidental because the attribute components get trimmed away. Once I include the following line:
```C#
Console.WriteLine(new JsonConverterAttribute(typeof(int)).ConverterType);
```
The compiler immediately warns. So it seems to me that using `JsonStringEnumConverter` (or any `JsonConverterFactory` for that matter) is simply not viable in source gen.

Reply:
> @vitek-karas pointed out that the above not warning is accidental because the attribute components get trimmed away.

Is there an issue logged for this? That seems like an analysis hole to me.

Reply:
> > @vitek-karas pointed out that the above not warning is accidental because the attribute components get trimmed away.
> 
> Is there an issue logged for this? That seems like an analysis hole to me.

Vitek can correct me if I'm wrong for this case since he looked at it, but in general we don't analyze code that wasn't needed. So if the warning is for some piece of code that was trimmed, we'd not look at it.

Reply:
This is NOT a bug - there is now reflection access to the `Flags` property in the repro, so the AOT compiler doesn't keep any metadata about the property (or even the fact that there is a property, it only keeps the accessors), that includes not having any attributes for it. And since it's not going to keep the attributes, it doesn't run any validation on it and thus doesn't warn.

Note that if this sample is modified to use `RequiresUnreferencedCode` (so that analyzer/linker will actually react to it), then both analyzer and linker would report the warning. Because they're not as precise in "trimming" as AOT compiler is.

So while this doesn't warn, I would argue that we don't want to rely on that behavior, since even defining what the behavior looks like is pretty difficult - and implementing it in illink and especially in the analyzer might be pretty tricky.

Reply:
So the "trick" here is that no preserved code accesses the `JsonConverterAttribute.ConverterType` property? That allows the AOT compiler to not warn?

Does the ctor that is mark as `RequiresDynamicCode` still get preserved? Even though no one accesses the `ConverterType` property?

Reply:
In the trimmer and AOT, if the .ctor is preserved we will report the RDC warning on it - always. The repro above actually does warn, but from the source generated code which calls the .ctor directly.

Analyzer currently doesn't correctly implement this behavior (combination of DAM and RDC) and will not report a warning (it will for DAM and RUC - don't ask ;-))